src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -3,45 +3,418 @@
+                                    break;
+
+                                case CorElementType.ELEMENT_TYPE_VAR:
+                                case CorElementType.ELEMENT_TYPE_MVAR:
+                                case CorElementType.ELEMENT_TYPE_STRING:
+                                case CorElementType.ELEMENT_TYPE_SZARRAY:
+                                case CorElementType.ELEMENT_TYPE_ARRAY:
+                                case CorElementType.ELEMENT_TYPE_CLASS:
+                                case CorElementType.ELEMENT_TYPE_OBJECT:
+                                    isGcField = true;
+                                    alignment = pointerSize;
+                                    size = pointerSize;
+                                    break;
+
+                                    // Statics for valuetypes where the valuetype is defined in this module are handled here. 
+                                    // Other valuetype statics utilize the pessimistic model below.
+                                case CorElementType.ELEMENT_TYPE_VALUETYPE:
+                                    isGcField = true;
+                                    alignment = pointerSize;
+                                    size = pointerSize;
+                                    break;
+
+                                case CorElementType.ELEMENT_TYPE_END:
+                                default:
+                                    isGcField = true;
+                                    alignment = pointerSize;
+                                    size = pointerSize;
+                                    if (!valueTypeHandle.IsNil)
+                                    {
+                                        // Allocate pessimistic non-GC area for cross-module fields as that's what CoreCLR does
+                                        // <a href="https://github.com/dotnet/coreclr/blob/659af58047a949ed50d11101708538d2e87f2568/src/vm/ceeload.cpp#L2124">here</a>
+                                        nonGcStatics[index] = LayoutInt.AlignUp(nonGcStatics[index], new LayoutInt(TargetDetails.MaximumPrimitiveSize))
+                                            + new LayoutInt(TargetDetails.MaximumPrimitiveSize);
+                                    }
+                                    else
+                                    {
+                                        // Field has an unexpected type
+                                        throw new InvalidProgramException();
+                                    }
+                                    break;
+                            }
+
+                            LayoutInt[] layout = (isGcField ? gcStatics : nonGcStatics);
+                            LayoutInt offset = LayoutInt.AlignUp(layout[index], new LayoutInt(alignment));
+                            layout[index] = offset + new LayoutInt(size);
+                            FieldDesc fieldDesc = null;
+                            try
```

Comment:
AFAIK `module.GetField` would pretty much only throw if there's an OOM (step through it in the debugger please to make sure I'm right). The try/catch is unnecessary.

Reply:
OK, removed.